### ISSUE TITLE
Porting into another language

### ISSUE THREAD

**Issue by lintangtimur** [lintangtimur/ovoid#11]
I hope someone want to porting into another language like
Javascript, Python, GO, Java

**Comment by anysz** [comment]
Interesting

**Comment by anysz** [comment]
### Okay, heads up! one more thing to test. I just need to logout OVO Account from my phone either register the new one, but everything is already ported

![image](https://user-images.githubusercontent.com/26342994/55275816-45ff6300-531e-11e9-9c8f-4600dc984e44.png)

**Comment by lintangtimur** [comment]
change deviceId or check logout function

**Comment by anysz** [comment]
> change deviceId or check logout function

Yes, in couple of minutes. I leave my phone on my house - I already logged into the OVO App on Android so i guess i need to logout from it?
Or, Have you managed to signup/register new phone number?

**Comment by lintangtimur** [comment]
2 Options
- [ ] from Meta.php there is deviceID, you must change that value https://github.com/lintangtimur/ovoid/issues/4#issuecomment-469552527
- [ ] logout from your phone or try using logout endpoint

"Have you managed to signup/register new phone number?"
not yet

**Comment by anysz** [comment]
What the maximum allowed of device logged in? (i didn't login with this api neither my prototype so there's no possiblity for unused login session)
1 device only? (the only one logged on my phone?)
Or i need file the support to logout entire login session? (because idk what device logged, i just remembered i have one and that's my phone)
Anw about the deviceId, it's random UUID right?
Also, what about [this one](https://github.com/lintangtimur/ovoid/blob/master/src/Ovoid.php#L84)? Why you hardcoded them?
(Sorry if i ask too many question)

**Comment by lintangtimur** [comment]
> 
> 
> What the maximum allowed of device logged in? (i didn't login with this api neither my prototype so there's no possiblity for unused login session)
> 1 device only? (the only one logged on my phone?)
> Or i need file the support to logout entire login session? (because idk what device logged, i just remembered i have one and that's my phone)
> Anw about the deviceId, it's random UUID right?
> Also, what about [this one](https://github.com/lintangtimur/ovoid/blob/master/src/Ovoid.php#L84)? Why you hardcoded them?
> (Sorry if i ask too many question)

from what i know, only 1 device only. Yes random UUID.
FCM(google push notification), i think it doesn't matter so i make it hardcoded

**Comment by anysz** [comment]
Ok most of them already ported now. But somehow i can't transfer the OVO Balance within this error. 
You sure the payload is right?
![image](https://user-images.githubusercontent.com/26342994/55293761-133f9280-5424-11e9-8791-248baf46427a.png)
![image](https://user-images.githubusercontent.com/26342994/55293718-bba12700-5423-11e9-804f-c2126925cd20.png)

**Comment by lintangtimur** [comment]
> 
> 
> Ok most of them already ported now. But somehow i can't transfer the OVO Balance within this error.
> You sure the payload is right?
> ![image](https://user-images.githubusercontent.com/26342994/55293761-133f9280-5424-11e9-8791-248baf46427a.png)
> ![image](https://user-images.githubusercontent.com/26342994/55293718-bba12700-5423-11e9-804f-c2126925cd20.png)

your error same like me, if i remember correctly payload for transfer endpoint is right, can't test it yet

**Comment by anysz** [comment]
Or this could be the account? Because OVO has 2, the one that after you registered, and one other after you file the picture of you and National ID Card (if I remembered, this one able to do balance transfer)

**Comment by lintangtimur** [comment]
my ovo app registered with national id card, and if you want to transfer you need to rgister first

**Comment by anysz** [comment]
Checkout ported Python3 one here, https://github.com/anysz/ovopy

**Comment by lintangtimur** [comment]
> 
> 
> Checkout ported Python3 one here, https://github.com/anysz/ovopy

thanks, great examples and documentation. Your repo already added into my ported list

**Comment by anysz** [comment]
Yea you're welcome 😉 I might be port them to JS too (since I mainly use JS for web development), how about doing payment such as purchasing Mobile data bandwidth, or Game Voucher or.... Pay electricity/telephone bills? This would be cool right! imagine you can just build your own payment billing website

**Comment by lintangtimur** [comment]
> 
> 
> Yea you're welcome 😉 I might be port them to JS too (since I mainly use JS for web development), how about doing payment such as purchasing Mobile data bandwidth, or Game Voucher or.... Pay electricity/telephone bills? This would be cool right! imagine you can just build your own payment billing website. _Industry 4.0 has begun_.

that's cool things, i will check again when i have much more free time, or you can do it your self

**Comment by anysz** [comment]
Any chance you implement that? :/ 

> Yea you're welcome  I might be port them to JS too (since I mainly use JS for web development), how about doing payment such as purchasing Mobile data bandwidth, or Game Voucher or.... Pay electricity/telephone bills? This would be cool right! imagine you can just build your own payment billing website. _Industry 4.0 has begun_.

**Comment by lintangtimur** [comment]
> 
> 
> Any chance you implement that? :/
> 
> > Yea you're welcome  I might be port them to JS too (since I mainly use JS for web development), how about doing payment such as purchasing Mobile data bandwidth, or Game Voucher or.... Pay electricity/telephone bills? This would be cool right! imagine you can just build your own payment billing website. _Industry 4.0 has begun_.

Ya, saya sedang mengerjakan bagian beli pulsa melalui OVO CASH, tapi sepertinya agak lama karena ada urusan lain, kalau anda ingin membantu silahkan. Chat via discord

**Comment by lintangtimur** [comment]
contoh respon
![image](https://user-images.githubusercontent.com/16686825/57027628-20d87a00-6c67-11e9-978a-fb90cfbea995.png)

**Comment by anysz** [comment]
Okei, take your time 😃

**Comment by rizkifani25** [comment]
about 'invalid model', already fixed or not? Still can't transfer OVO Balance

**Comment by anysz** [comment]
Awaiting confirmation @lintangtimur

**Comment by lintangtimur** [comment]
> 
> 
> about 'invalid model', already fixed or not? Still can't transfer OVO Balance

I hope it fixed ee0745a73766491f624b2ab2bcebd2b92e0dcee0, check latest released

**Comment by apriady** [comment]
> Yea you're welcome 😉 I might be port them to JS too (since I mainly use JS for web development), how about doing payment such as purchasing Mobile data bandwidth, or Game Voucher or.... Pay electricity/telephone bills? This would be cool right! imagine you can just build your own payment billing website. _Industry 4.0 has begun_.

Hi, have you start on JS?
I'm interested in porting it to nodejs, maybe we can port it together, dude?

**Comment by anysz** [comment]
> > Yea you're welcome wink I might be port them to JS too (since I mainly use JS for web development), how about doing payment such as purchasing Mobile data bandwidth, or Game Voucher or.... Pay electricity/telephone bills? This would be cool right! imagine you can just build your own payment billing website. _Industry 4.0 has begun_.
> 
> Hi, have you start on JS?
> I'm interested in porting it to nodejs, maybe we can port it together, dude?

I've tried but I discontinued my work (during last year of study) and leave it with no update, all goes back to you :/ you could invite me to the repository soon as you start porting to nodejs 😉

**Comment by apriady** [comment]
@lintangtimur NodeJS https://github.com/apriady/ovoid-nodejs

@anysz I've created that repo. has ported several method to NodeJS but still full of messiness and not structured well. Maybe can give suggestion to make it better. 
Thanks

**Comment by anysz** [comment]
@apriady whoa that's cool, however regarding to this issue https://github.com/lintangtimur/ovoid/issues/11#issuecomment-478371791 I think it's a must to store Auth token in case you forgot to save it and throw maximum device  next time if you trying to login, you could take a look to https://github.com/anysz/ovopy for the implementation probably

**Comment by apriady** [comment]
@anysz yes, i think that must implemented. I will look into your code and see how it works. Thanks for the suggestion.

BTW, anyone know the endpoint and params that used for refreshing the token?
i think it will be better if we can implement that.

**Comment by apriady** [comment]
halo @lintangtimur, can share tutorial that you use to setup the ssl pinning?
I'm following this https://omespino.com/tutorial-universal-android-ssl-pinning-in-10-minutes-with-frida/ but no luck.

**Comment by lintangtimur** [comment]
> 
> 
> halo @lintangtimur, can share tutorial that you use to setup the ssl pinning?
> I'm following this https://omespino.com/tutorial-universal-android-ssl-pinning-in-10-minutes-with-frida/ but no luck.

U can use xposed framework(root needed), or use Frida+Objection+Burp suite+Android virtual device emulator

**Comment by adibaulia** [comment]
Halo @lintangtimur, ini saya mau mencoba porting ke Go language, cuman waktu nyoba ambil auth Token (third step) ada error kayak gini.
```
 {
    "message": "Invalid parameters",
    "code": 1000
}
```
Mungkin ada solusi? oiya apa udah ada dokumentasi API nya dari OVO ya di postman gitu?

**Comment by lintangtimur** [comment]
> 
> 
> Halo @lintangtimur, ini saya mau mencoba porting ke Go language, cuman waktu nyoba ambil auth Token (third step) ada error kayak gini.
> 
> ```
>  {
>     "message": "Invalid parameters",
>     "code": 1000
> }
> ```
> 
> Mungkin ada solusi? oiya apa udah ada dokumentasi API nya dari OVO ya di postman gitu?

saya test masih bisa, lancar. Cek lagi parameternya

**Comment by adibaulia** [comment]
> > Halo @lintangtimur, ini saya mau mencoba porting ke Go language, cuman waktu nyoba ambil auth Token (third step) ada error kayak gini.
> > ```
> >  {
> >     "message": "Invalid parameters",
> >     "code": 1000
> > }
> > ```
> > 
> > 
> > Mungkin ada solusi? oiya apa udah ada dokumentasi API nya dari OVO ya di postman gitu?
> 
> saya test masih bisa, lancar. Cek lagi parameternya

udah saya test sih ini, jadi saya lagi bikin dokumentasi Apinya di postman gitu baca dari source codenya langsung.

Headernya pakai ini kan?
`Content-Type:application/json
app-id:C7UMRSMFRZ46D9GW9IK7
App-Version:3.2.0
OS:Android`
lalu paramnya pakai json ini
`{
	"deviceUnixtime":1543693061,
	"securityCode":"21xx",
	"updateAccessToken":"f6ec49c6f96549598c04b482ea49caxxx",
	"message":""
}`

**Comment by lintangtimur** [comment]
> 
> 
> > > Halo @lintangtimur, ini saya mau mencoba porting ke Go language, cuman waktu nyoba ambil auth Token (third step) ada error kayak gini.
> > > ```
> > >  {
> > >     "message": "Invalid parameters",
> > >     "code": 1000
> > > }
> > > ```
> > > 
> > > 
> > > Mungkin ada solusi? oiya apa udah ada dokumentasi API nya dari OVO ya di postman gitu?
> > 
> > 
> > saya test masih bisa, lancar. Cek lagi parameternya
> 
> udah saya test sih ini, jadi saya lagi bikin dokumentasi Apinya di postman gitu baca dari source codenya langsung.
> 
> Headernya pakai ini kan?
> 
> ```
> Content-Type:application/json
> app-id:C7UMRSMFRZ46D9GW9IK7
> App-Version:3.2.0
> OS:Android
> ```
> 
> lalu paramnya pakai json ini?
> 
> ```
> {
> 	"deviceUnixtime":1543693061,
> 	"securityCode":"21xx",
> 	"updateAccessToken":"f6ec49c6f96549598c04b482ea49caxxx",
> 	"message":""
> }
> ```

benar, itu security code bukannya 6digit?

**Comment by adibaulia** [comment]
> > > > Halo @lintangtimur, ini saya mau mencoba porting ke Go language, cuman waktu nyoba ambil auth Token (third step) ada error kayak gini.
> > > > ```
> > > >  {
> > > >     "message": "Invalid parameters",
> > > >     "code": 1000
> > > > }
> > > > ```
> > > > 
> > > > 
> > > > Mungkin ada solusi? oiya apa udah ada dokumentasi API nya dari OVO ya di postman gitu?
> > > 
> > > 
> > > saya test masih bisa, lancar. Cek lagi parameternya
> > 
> > 
> > udah saya test sih ini, jadi saya lagi bikin dokumentasi Apinya di postman gitu baca dari source codenya langsung.
> > Headernya pakai ini kan?
> > ```
> > Content-Type:application/json
> > app-id:C7UMRSMFRZ46D9GW9IK7
> > App-Version:3.2.0
> > OS:Android
> > ```
> > 
> > 
> > lalu paramnya pakai json ini?
> > ```
> > {
> > 	"deviceUnixtime":1543693061,
> > 	"securityCode":"21xx",
> > 	"updateAccessToken":"f6ec49c6f96549598c04b482ea49caxxx",
> > 	"message":""
> > }
> > ```
> 
> benar, itu security code bukannya 6digit?

oh oke, lagi dicoba porting nih mas, tapi belum selesai. sila coba ke [ovoid-go](https://github.com/adibaulia/ovoid-go)

**Comment by bintang007** [comment]
Izin porting ke java

**Comment by namdevel** [comment]
@lintangtimur Ruby https://github.com/namdevel/ovoid-ruby

**Comment by lintangtimur** [comment]
> 
> 
> @lintangtimur Ruby https://github.com/namdevel/ovoid-ruby

Done, ty :)

**Comment by namdevel** [comment]
nggih mas

**Comment by maulana20** [comment]
hallo mas @lintangtimur boleh saya izin porting ke dart, sebagai coba coba :blush:

**Comment by lintangtimur** [comment]
> hallo mas @lintangtimur boleh saya izin porting ke dart, sebagai coba coba 😊

silahkan, bebas

**Comment by maulana20** [comment]
> > hallo mas @lintangtimur boleh saya izin porting ke dart, sebagai coba coba 😊
> 
> silahkan, bebas

beberapa ada yang sudah bisa mas, silahkan coba [ovoid-flutter](https://github.com/maulana20/ovoid-flutter)

**Comment by robintan** [comment]
do you still need help in porting? I can help porting this completely into nodejs / python
and I am also looking to hire someone who understand reverse engineering / intercepting traffic like you. please contact me via whatsapp at +6281282888200

**Comment by AlfinAI** [comment]
Halo mas @lintangtimur, izin share di thread ini 🙏

Saya sudah porting ovoid ke Python: **[AlfinAI/ovoid-python](https://github.com/AlfinAI/ovoid-python)** (PyPI: `pip install ovoid`).

- Port lengkap: 14 services — login OTP + security code, balance, transactions, dll.
- Diverifikasi silang dengan teardown APK OVO asli + catatan riset di `research/`
- Semua tes offline (mock transport), 100% coverage, full docstring + type hints
- Atribusi penuh ke repo ini di README, LICENSE, pyproject, dan release notes

Terima kasih banyak untuk riset aslinya — repo ini yang bikin port ini memungkinkan. Kalau berkenan, silakan ditautkan 🙏
